alpheios_nemo_ui/data/assets/static/sw.js

```diff
@@ -0,0 +1,87 @@
+/* global self, workbox, Response, URL */
+console.log('Service worker has been registered')
+
+self.importScripts('/workbox/workbox-sw.js');
+
+self.addEventListener('install', event => {
+  console.log(`Service worker install event`)
+})
+
+self.addEventListener('activate', event => {
+  console.log(`Service worker activate event`)
+})
+
+// This code runs whenever a Service Worker script is loaded, and Workbox library is loaded too
+if (workbox) {
+  console.log(`workbox is active`)
+  // workbox.core.setLogLevel(workbox.core.LOG_LEVELS.debug)
+
+  self.addEventListener('fetch', evt => {
+    console.log(`Service worker fetch evt: ${evt.request.url}`, evt)
+  })
+
+  // Will it cause an error if overwrite current cache files as with Cache.addAll()?
+  self.__precacheManifest = [].concat(self.__precacheManifest || [])
+  workbox.precaching.precacheAndRoute(
+    self.__precacheManifest
+  )
+
+  workbox.routing.registerRoute(
+    // Cache image files
+    /.*\.(?:png|jpg|jpeg|svg|gif)/,
+    // Use the cache if it's available
+    new workbox.strategies.StaleWhileRevalidate({
+      // Use a custom cache name
+      cacheName: 'image-cache',
+      plugins: [
+        new workbox.expiration.Plugin({
+          // Cache only 20 images
+          maxEntries: 20,
+          // Cache for a maximum of a week
+          maxAgeSeconds: 7 * 24 * 60 * 60
+        })
+      ]
+    })
+  )
+
+  // External resources
+  workbox.routing.registerRoute(
+    /(?:https?:\/\/).*/,
```

Comment:
I think I would prefer that we start off with explicit strategies that differentiate between our application's static resources and those from external sources (such as googleapis for the fonts, etc)

Reply:
> is this always going to catch the local javascript and css too?

Yes, I think it will. Any local resource will be visible as `http(s)://server_url/path/resource_file`, so it will match this regexp's test condition. Local files were served from the service worker's cache in my tests:
![image](https://user-images.githubusercontent.com/18631055/59209103-9ff09480-8bbb-11e9-98f9-8eeb535fb384.png)

I think the comment is a bit misleading. It was assuming that all local resources will be precached in earlier statements so all files caught by this statement be external. We're not doing that in Alpheios UI so this condition will catch both local and external files.

Reply:
> I think I would prefer that we start off with explicit strategies that differentiate between our application's static resources and those from external sources (such as googleapis for the fonts, etc)

The standard practice is to pre-cache all local files. With one or several webpack plugins each local resource has a content hash added to its name, so that when the local resource is updated, it will have a new name, and will be loaded as a new file, no matter what cache policy is used. A webpack plugin then generates a list of local resources and adds a command to pre-cache them to the index HTML file, usually. Every time any of local resources is updated, a webpack built process is rerun and a list of resources is updated.

This is done to guarantee that all local resources loaded every single moment of time are of the same "generation" (i.e. we do not have a.css version 1 loaded together with b.css version 2). 

To implement this in a Flask app is a more complex task than in a regular webapp. A webapp relies on JS for templating, and an `index.html` is the file that needs to be updated by the webpack build process. The webpack plugins are available to handle this. In the Flask framework, it's usually one or several templates that need to be updated instead. I'm not sure there are any webpack plugins that can handle that, but maybe we can deal with it somehow still. So should we try to implement a process of pre-caching for the Nemo UI? What do you think?